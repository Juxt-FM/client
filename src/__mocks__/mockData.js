import faker from "faker";

export const getMockUser = () => ({
  id: "1",
  email: {
    address: faker.internet.email(),
    verified: true,
  },
  phone: {
    number: faker.phone.phoneNumber(),
    verified: true,
  },
  profile: {
    id: "1",
    name: faker.name.findName(),
    location: faker.address.city + ", " + faker.address.stateAbbr(),
    summary: faker.company.catchPhrase(),
    imageURL: faker.image.imageUrl(),
    watchlists: [],
  },
  verified: true,
  active: true,
  suspended: false,
  lastLogin: new Date(),
  updatedAt: new Date(),
  createdAt: new Date(),
});

export const getMockPost = () => ({
  id: "1",
  publicationStatus: "public",
  contentFormat: "markdown",
  author: "1",
  title: "Is the market catching up with the economy?",
  subtitle:
    "Lorem ipsum dolor sit, amet consectetur adipisicing elit. Consectetur inventore rem quaerat facere sit, est nulla sed illum molestias dicta ipsa corporis ipsam porro dolore eaque architecto velit voluptates quam.",
  imageURL: faker.image.imageUrl(),
  content:
    "Lorem ipsum dolor sit, amet consectetur adipisicing elit. Consectetur inventore rem quaerat facere sit, est nulla sed illum molestias dicta ipsa corporis ipsam porro dolore eaque architecto velit voluptates quam.",
  symbols: ["NVDA", "QQQ"],
  tags: ["stock market"],
  comments: [],
  reactionCount: 0,
  reactionStatus: null,
  createdAt: "1609986720",
  updatedAt: "1609986720",
});
