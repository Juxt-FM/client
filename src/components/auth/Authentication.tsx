/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import Link from "next/link";

import styles from "../styles/modules/auth.module.scss";

interface IHeader {
  title: string;
  description?: string;
  color: "black" | "accent";
}

export const Header = ({ title, description, color }: IHeader) => {
  return (
    <div className={styles.header}>
      <Link href="/">
        <a>
          <img
            src={require(`../images/logo-${color}-lg.png`)}
            alt="brand logo"
            className={styles.logo}
          />
        </a>
      </Link>
      <h2>{title}</h2>
      {description && <p>{description}</p>}
    </div>
  );
};

interface IContentSection {
  title: string;
  content: string;
}

export const ContentSection = ({ title, content }: IContentSection) => (
  <section className={styles.content}>
    <h2>{title}</h2>
    <p>{content}</p>
  </section>
);
