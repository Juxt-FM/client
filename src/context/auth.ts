/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { useEffect } from "react";
import { useApolloClient, useLazyQuery } from "@apollo/client";
import { useSelector, useDispatch } from "react-redux";
import Cookies from "universal-cookie";
import jwt_decode from "jwt-decode";

import {
  FETCH_USER,
  FETCH_USER_FAIL,
  FETCH_USER_SUCCESS,
  LOGIN_USER,
  LOGOUT_USER,
  REFRESH_TOKEN,
  REFRESH_TOKEN_FAIL,
  REFRESH_TOKEN_SUCCESS,
  RootState,
} from "../store";

import {
  QUERY_AUTH_USER,
  MUTATION_REFRESH_TOKEN,
  User,
  setAccessToken,
  lockApollo,
  unlockApollo,
  createApolloClient,
  AuthCredentials,
} from "../graphql";
import { usePreviousValue } from "./utils";
import { useRouter } from "next/router";

const cookies = new Cookies();

const buildTokenInfo = (token: string) => {
  const decoded: any = jwt_decode(token);
  return {
    expires: decoded.exp,
    issued: decoded.iat,
  };
};

const refreshToken = () => {
  const client = createApolloClient({ standalone: true });

  // eslint-disable-next-line no-async-promise-executor
  return new Promise(async (resolve, reject) => {
    try {
      const { data } = await client.mutate({
        mutation: MUTATION_REFRESH_TOKEN,
      });
      if (data) {
        resolve(data.refreshToken);
      } else {
        reject(new Error());
      }
    } catch (e) {
      reject(e);
    }
  });
};

// Returns whether or not the user is logged in
export const useAuthStatus = () => cookies.get("logged_in") === "true";

interface IProvider {
  children: JSX.Element;
}

export const AuthProvider = ({ children }: IProvider): JSX.Element => {
  const loggedIn = useAuthStatus();
  const { onLogout } = useAuthActions();

  const { token } = useSelector((state: RootState) => state.auth);

  const prevToken = usePreviousValue(token);

  const dispatch = useDispatch();

  const onRefreshTokenSuccess = ({ accessToken }: AuthCredentials) => {
    dispatch({
      type: REFRESH_TOKEN_SUCCESS,
      payload: buildTokenInfo(accessToken),
    });

    setAccessToken(accessToken);
    unlockApollo();
  };

  const onRefreshTokenFailed = () => {
    dispatch({ type: REFRESH_TOKEN_FAIL });

    unlockApollo();

    onLogout(false);
  };

  const refresh = () => {
    lockApollo();
    dispatch({ type: REFRESH_TOKEN });
    refreshToken().then(onRefreshTokenSuccess).catch(onRefreshTokenFailed);
  };

  useEffect(() => {
    if (loggedIn) refresh();
  }, []);

  /**
   * Refresh the user's token every 14:45 minutes.
   * need to make sure the token info is actually new
   */
  useEffect(() => {
    if (token && token !== prevToken) {
      const timeout = token.expires * 1000 - token.issued * 1000 - 15000;

      // tslint:disable-next-line:no-console
      console.log(`Refreshing token in ${timeout / 1000 / 60} minutes`);

      setTimeout(refresh, timeout);
    }
  }, [token]);

  return children;
};

export function useAuthActions() {
  const dispatch = useDispatch();
  const client = useApolloClient();
  const router = useRouter();

  const onLogin = (accessToken: string) => {
    dispatch({ type: LOGIN_USER, payload: buildTokenInfo(accessToken) });

    cookies.set("logged_in", "true", { path: "/" });
    setAccessToken(accessToken);

    client.clearStore().then(() => router.push("/"));
  };

  const onLogout = (redirectToLogin = true) => {
    dispatch({ type: LOGOUT_USER });

    setAccessToken(undefined);
    cookies.set("logged_in", undefined, { path: "/", expires: new Date() });

    client.clearStore().then(() => {
      if (redirectToLogin) router.push("/auth/login");
    });
  };

  return {
    onLogin,
    onLogout,
  };
}

export const UserProvider = ({ children }: IProvider) => {
  const loggedIn = useAuthStatus();

  const dispatch = useDispatch();

  const [fetchUser, { data, called }] = useLazyQuery<{
    me: User;
  }>(QUERY_AUTH_USER, {
    onError: (err) => {
      dispatch({ type: FETCH_USER_FAIL, payload: err });
    },
  });

  useEffect(() => {
    if (loggedIn) fetchUser();
  }, [loggedIn]);

  useEffect(() => {
    if (called) dispatch({ type: FETCH_USER });
  }, [called]);

  useEffect(() => {
    if (data) {
      dispatch({ type: FETCH_USER_SUCCESS, payload: data.me });
    }
  }, [data]);

  return children;
};

export const useAuthUser = () =>
  useSelector((state: RootState) => state.auth.user);
