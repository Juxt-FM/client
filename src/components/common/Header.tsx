/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import React from "react";
import { useState, MouseEvent } from "react";
import { useRouter } from "next/router";
import Link from "next/link";
import { useCurrentBreakpoint } from "../../context";
import { Squash as Hamburger } from "hamburger-react";
import Searchbar from "./Searchbar";

import styles from "../../styles/modules/header.module.scss";

const Logo = () => {
  const router = useRouter();

  const onClick = (e: MouseEvent<HTMLAnchorElement, globalThis.MouseEvent>) => {
    e.preventDefault();
    router.push("/");
  };

  return (
    <a href="/" onClick={onClick}>
      <img
        className={styles.headerLogo}
        src={require("../../images/logo-black-sm.png")}
        alt="brand_logo"
      />
    </a>
  );
};

interface INavLink {
  label: string;
  path: string;
  className?: string;
}

const NavLink = ({ label, path, className }: INavLink) => (
  <Link href={path}>
    <a className={className}>{label}</a>
  </Link>
);

export const ContentHeader = ({ searchEnabled = true }) => {
  const currentBreakpoint = useCurrentBreakpoint();
  const [showMenu, setShowMenu] = useState(false);

  const renderMenuAction = () => {
    return (
      <div className={styles.menuAction}>
        <Hamburger
          size={24}
          color="black"
          toggled={showMenu}
          toggle={setShowMenu}
        />
      </div>
    );
  };

  return (
    <header className={styles.header}>
      <div className={styles.headerLeft}>
        <Logo />
        {searchEnabled && (
          <Searchbar placeholder="Search blog posts, stocks, and more..." />
        )}
      </div>
      {["xs", "sm", "md"].includes(currentBreakpoint) && renderMenuAction()}
      <ul className={styles.headerLinks}>
        <li className={styles.headerLink}>
          <NavLink label="Home" path="/" />
        </li>
        <li className={styles.headerLink}>
          <NavLink label="Stocks" path="/stocks" />
        </li>
        <li className={styles.headerLink}>
          <NavLink label="Cryptocurrencies" path="/crypto" />
        </li>
      </ul>
    </header>
  );
};

interface IRootProps {
  children: JSX.Element;
}

const HeaderRoot = ({ children }: IRootProps) => {
  return <div className={styles.root}>{children}</div>;
};

export default HeaderRoot;
