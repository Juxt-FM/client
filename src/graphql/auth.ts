/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { setContext } from "@apollo/client/link/context";

// A user's access token, expires every 15 minutes
let accessToken: string | undefined;

let locked = false;

/**
 * Updates the access token in memory
 * @param newToken
 */
export const setAccessToken = (newToken: string | undefined) => {
  if (newToken !== accessToken) {
    // tslint:disable-next-line:no-console
    console.log("updating token");

    accessToken = newToken;
  }
};

/**
 * Lock the apollo client to make
 * all requests wait until it is unlocked
 */
export const lockApollo = () => {
  // tslint:disable-next-line:no-console
  console.log("locking");

  locked = true;
};

/**
 * Unlock apollo client, will initiate
 * all waiting requests
 */
export const unlockApollo = () => {
  // tslint:disable-next-line:no-console
  console.log("unlocking");

  locked = false;
};

/**
 * This injects the user's access token into the request headers,
 * this happens for every request
 *
 * We check to see if the client is unlocked before resolving the new headers.
 * If locked, use setTimeout to try again in 100ms. Else, resolve the headers
 */
export const authLink = setContext((_: any, { headers }: any) => {
  const retries = 0;
  const promise = () =>
    new Promise<any>((resolve) => {
      const buildContext = () => {
        if (!locked) {
          // tslint:disable-next-line:no-console
          console.log("apollo unlocked, fetching");

          if (accessToken) {
            resolve({
              headers: {
                ...headers,
                Authorization: `Bearer ${accessToken}`,
              },
            });
          } else {
            resolve(headers);
          }

          return;
        } else if (retries > 15) {
          return headers;
        } else {
          // tslint:disable-next-line:no-console
          console.log("apollo locked, waiting");

          setTimeout(buildContext, 100);
        }
      };

      buildContext();
    });

  return promise().then((authenticatedHeaders) => authenticatedHeaders);
});
