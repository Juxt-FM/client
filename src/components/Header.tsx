/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import React, { Fragment, useState } from "react";
import { useAuthStatus, useAuthUser } from "../lib/context";

import {
  faArrowLeft,
  faCog,
  faCommentAlt,
  faEllipsisH,
  faQuestionCircle,
  faSearch,
  faShieldAlt,
  faSignOutAlt,
  faUser,
} from "@fortawesome/free-solid-svg-icons";
import { Squash as Hamburger } from "hamburger-react";

import styles from "../styles/modules/header.module.scss";
import Dropdown, { DropdownList } from "./common/Dropdown";
import { useRouter } from "next/router";
import { IconAction } from "./common/Buttons";
import Link from "next/link";

interface IPageInfo {
  title: string;
  backButton?: boolean;
}

const PageInfo = ({ backButton = false, title }: IPageInfo) => {
  const router = useRouter();
  const [showMenu, setShowMenu] = useState(false);

  const renderActionButton = () => {
    if (showMenu && !backButton) {
      return (
        <div className={styles.menuAction}>
          <Hamburger
            size={24}
            color="black"
            toggled={showMenu}
            toggle={setShowMenu}
          />
        </div>
      );
    } else if (backButton) {
      return (
        <div className={styles.backButton}>
          <IconAction icon={faArrowLeft} onClick={() => router.back()} />
        </div>
      );
    }
  };

  return (
    <div className={styles.pageInfo}>
      {renderActionButton()}
      <h3 className={styles.title}>{title}</h3>
    </div>
  );
};

const AccountDropdown = () => {
  const router = useRouter();
  const { user } = useAuthUser();

  const renderAnchor = (
    openDropdown: (
      event: React.MouseEvent<HTMLAnchorElement, MouseEvent>
    ) => void
  ) => (
    <a className={styles.navOption} onClick={openDropdown}>
      <img
        className={styles.profileImage}
        src="https://www.pererasys.com/_next/static/images/me-0e70979b96b772f832a278693ee0cd0e.jpg"
        alt="logged in user"
      />
    </a>
  );

  const renderContent = () => {
    const lists = [
      [
        {
          label: "Profile",
          icon: faUser,
          action: () => router.push(`/users/${user.id}`),
        },
      ],
      [
        {
          label: "Settings",
          icon: faCog,
          action: () => router.push(`/settings/account`),
        },
        {
          label: "Privacy",
          icon: faShieldAlt,
          action: () => router.push(`/settings/privacy`),
        },
      ],
      [
        {
          label: "Help Center",
          icon: faQuestionCircle,
          action: () => router.push(`/support/help`),
        },
        {
          label: "Contact Us",
          icon: faCommentAlt,
          action: () => router.push(`/support/contact`),
        },
      ],
      [
        {
          label: "Logout",
          icon: faSignOutAlt,
          action: () => router.push("/auth/logout"),
          danger: true,
        },
      ],
    ];

    return (
      <Fragment>
        {lists.map((options, index) => (
          <Fragment key={String(index)}>
            <DropdownList options={options} />
            {index + 1 !== lists.length && <hr />}
          </Fragment>
        ))}
      </Fragment>
    );
  };

  if (user) {
    return <Dropdown anchor={renderAnchor} content={renderContent} />;
  } else {
    return <div className={styles.loadingNavOption}></div>;
  }
};

const Navigation = () => {
  const loggedIn = useAuthStatus();
  return (
    <ul className={styles.navigation}>
      <li>
        <IconAction icon={faSearch} onClick={() => {}} />
      </li>
      <li>
        <IconAction icon={faEllipsisH} onClick={() => {}} />
      </li>
      {loggedIn && (
        <li>
          <AccountDropdown />
        </li>
      )}
    </ul>
  );
};

interface IHeader extends IPageInfo {}

export const Header = ({ title, backButton }: IHeader) => {
  return (
    <header className={styles.header}>
      <PageInfo title={title} backButton={backButton} />
      <Navigation />
    </header>
  );
};

export default Header;
