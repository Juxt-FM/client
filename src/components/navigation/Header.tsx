/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import React, { Fragment, useState } from "react";
import { useRouter } from "next/router";
import { useAuthStatus, useAuthUser } from "../../lib/auth";

import {
  faArrowLeft,
  faCog,
  faCommentAlt,
  faEllipsisH,
  faQuestionCircle,
  faSearch,
  faShieldAlt,
  faSignOutAlt,
  faUser,
} from "@fortawesome/free-solid-svg-icons";
import { Squash as Hamburger } from "hamburger-react";

import Dropdown, { DropdownList } from "../common/Dropdown";
import { Button, IconAction } from "../common/Buttons";

import styles from "../../styles/navigation/header.module.scss";

interface IPageInfo {
  title: string;
  backButton?: boolean;
}

export const BackButton = () => {
  const router = useRouter();

  const onBack = () => router.back();

  return (
    <div className={styles.backButton}>
      <IconAction icon={faArrowLeft} size="sm" onClick={onBack} />
    </div>
  );
};

export const PageInfo = ({ backButton = false, title }: IPageInfo) => {
  const [showMenu, setShowMenu] = useState(false);

  const renderActionButton = () => {
    if (showMenu && !backButton) {
      return (
        <div className={styles.menuAction}>
          <Hamburger
            size={24}
            color="black"
            toggled={showMenu}
            toggle={setShowMenu}
          />
        </div>
      );
    } else if (backButton) {
      return <BackButton />;
    }
  };

  return (
    <div className={styles.pageInfo}>
      {renderActionButton()}
      <h3 className={styles.title}>{title}</h3>
    </div>
  );
};

export const SettingsDropdown = () => {
  const renderAnchor = (
    openDropdown: (
      event: React.MouseEvent<HTMLAnchorElement, MouseEvent>
    ) => void
  ) => <IconAction icon={faEllipsisH} size="sm" onClick={openDropdown} />;

  const renderContent = () => {
    const lists = [
      [
        {
          label: "Settings",
          icon: faCog,
          path: `/app/settings/account`,
        },
        {
          label: "Privacy",
          icon: faShieldAlt,
          path: `/app/settings/privacy`,
        },
      ],
      [
        {
          label: "Help Center",
          icon: faQuestionCircle,
          path: `/support/help`,
        },
        {
          label: "Contact Us",
          icon: faCommentAlt,
          path: `/support/contact`,
        },
      ],
    ];

    return (
      <Fragment>
        {lists.map((options, index) => (
          <Fragment key={String(index)}>
            <DropdownList options={options} />
            {index + 1 !== lists.length && <hr />}
          </Fragment>
        ))}
      </Fragment>
    );
  };

  return <Dropdown renderAnchor={renderAnchor}>{renderContent()}</Dropdown>;
};

export const AccountDropdown = () => {
  const { user } = useAuthUser();

  const renderAnchor = (
    openDropdown: (
      event: React.MouseEvent<HTMLAnchorElement, MouseEvent>
    ) => void
  ) => (
    <a className={styles.navOption} onClick={openDropdown}>
      <img
        className={styles.profileImage}
        src="https://www.pererasys.com/_next/static/images/me-0e70979b96b772f832a278693ee0cd0e.jpg"
        alt="logged in user"
      />
    </a>
  );

  const renderContent = () => {
    const lists = [
      [
        {
          label: "Profile",
          icon: faUser,
          path: `/app/users/${user.id}`,
        },
      ],
      [
        {
          label: "Logout",
          icon: faSignOutAlt,
          path: "/auth/logout",
          danger: true,
        },
      ],
    ];

    return (
      <Fragment>
        {lists.map((options, index) => (
          <Fragment key={String(index)}>
            <DropdownList options={options} />
            {index + 1 !== lists.length && <hr />}
          </Fragment>
        ))}
      </Fragment>
    );
  };

  if (user) {
    return <Dropdown renderAnchor={renderAnchor}>{renderContent()}</Dropdown>;
  } else {
    return <div className={styles.loadingNavOption}></div>;
  }
};

export const Navigation = () => {
  const router = useRouter();
  const loggedIn = useAuthStatus();

  return (
    <ul className={styles.navigation}>
      <li className={styles.navItem}>
        <IconAction icon={faSearch} size="sm" onClick={() => {}} />
      </li>
      <li className={styles.navItem}>
        <SettingsDropdown />
      </li>
      <li className={styles.navItem}>
        {loggedIn ? (
          <AccountDropdown />
        ) : (
          <Button
            label="Get started"
            color="lightGreen"
            size="sm"
            onClick={() => router.push("/auth/signup")}
          />
        )}
      </li>
    </ul>
  );
};

const Header = ({ title, backButton }: IPageInfo) => {
  return (
    <header className={styles.header}>
      <PageInfo title={title} backButton={backButton} />
      <Navigation />
    </header>
  );
};

export default Header;
