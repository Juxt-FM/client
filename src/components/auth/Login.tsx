/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { ChangeEvent, FormEvent, useState } from "react";
import { ApolloError, useMutation } from "@apollo/client";
import Link from "next/link";

import { AuthCredentials, MUTATION_LOGIN } from "../../lib/graphql";
import { useAuthActions } from "../../lib/context";

import { ContentSection, Header } from "./Authentication";

import { FormInput } from "../common/Inputs";
import { Button } from "../common/Buttons";

import styles from "../../styles/auth/login.module.scss";

export const Error = ({ message }: { message: string }) => (
  <div className={styles.error}>
    <p>{message}</p>
  </div>
);

export const LoginErrors = ({ error }: { error: ApolloError }) => {
  const renderContent = () => {
    if (error) {
      if (error.graphQLErrors && error.graphQLErrors.length > 0) {
        return error.graphQLErrors.map((err) => (
          <Error message={err.message} key={err.message} />
        ));
      } else if (error.networkError) {
        return <Error message={error.networkError.message} />;
      }
    }
  };

  return <div className={styles.loginErrors}>{renderContent()}</div>;
};

const LoginForm = () => {
  const { onLogin } = useAuthActions();

  const [identifier, setIdentifier] = useState("");
  const [password, setPassword] = useState("");

  const [login, { loading, error }] = useMutation<{
    loginUser: AuthCredentials;
  }>(MUTATION_LOGIN, {
    variables: {
      data: {
        identifier,
        password,
      },
    },
    onCompleted: ({ loginUser: { accessToken } }) => {
      onLogin(accessToken);
    },
    // tslint:disable-next-line:no-empty
    onError: () => {},
  });

  const onIdentifierChange = (event: ChangeEvent<HTMLInputElement>) => {
    setIdentifier(event.target.value);
  };

  const onPasswordChange = (event: ChangeEvent<HTMLInputElement>) => {
    setPassword(event.target.value);
  };

  const onLoginUser = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    login();
  };

  return (
    <form id="loginForm" className={styles.loginForm} onSubmit={onLoginUser}>
      <FormInput
        id="identifier"
        name="identifier"
        onChange={onIdentifierChange}
        type="text"
        label="Email or phone number"
        required
      />
      <FormInput
        id="password"
        name="password"
        onChange={onPasswordChange}
        type="password"
        label="Password"
        required
      />
      <Button
        id="login"
        label="Log in"
        disabled={!identifier || !password}
        loading={loading}
        loadingLabel="Just a moment..."
        type="submit"
      />
      <LoginErrors error={error} />
    </form>
  );
};

export default function Login() {
  return (
    <div className={[styles.root, styles.reverse].join(" ")}>
      <div className={styles.leftLogin}>
        <ContentSection
          title="Title 1"
          content="
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores
                praesentium similique aliquam culpa suscipit, itaque, id architecto
                nihil voluptatem consequatur et atque aspernatur ex totam cupiditate
                obcaecati quos autem sed."
        />
        <ContentSection
          title="Title 2"
          content="
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Maiores
                praesentium similique aliquam culpa suscipit, itaque, id architecto
                nihil voluptatem consequatur et atque aspernatur ex totam cupiditate
                obcaecati quos autem sed."
        />
      </div>
      <div className={styles.rightLogin}>
        <div className={styles.loginWrapper}>
          <Header title="Log in to your account" color="black" />
          <LoginForm />
          <div className={styles.options}>
            <div className={styles.switchAction}>
              <p>Don't have an account?</p>
              <Link href="/auth/signup">
                <a>Sign up</a>
              </Link>
            </div>
            <Link href="/auth/forgot-password">
              <a>Forgot password</a>
            </Link>
          </div>
        </div>
      </div>
    </div>
  );
}
