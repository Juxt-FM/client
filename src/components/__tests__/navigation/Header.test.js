/**
 * @author Andrew Perera
 * Copyright (C) 2020 - All rights reserved
 */

import { create } from "react-test-renderer";

import Header, {
  AccountDropdown,
  BackButton,
  SettingsDropdown,
} from "../../navigation/Header";

import { useAuthStatus, useAuthUser } from "../../../lib/auth";
import { getMockUser } from "../../../__mocks__/mockData";
import { useRouter } from "next/router";

jest.mock("../../../lib/auth");
jest.mock("next/router");

useAuthUser.mockImplementation(() => ({
  user: getMockUser(),
  loading: false,
  error: undefined,
}));

const push = jest.fn();
const pathname = "/";

useRouter.mockImplementation(() => ({ push, pathname }));

describe("<Header />", () => {
  const setup = (props) => {
    const component = create(<Header {...props} />);

    return { component };
  };

  it("should render for logged in user without crashing", () => {
    useAuthStatus.mockImplementationOnce(() => "1");

    const props = {
      backButton: false,
      title: "Test Title",
    };

    const { component } = setup(props);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render for unauthenticated user without crashing", () => {
    useAuthStatus.mockImplementationOnce(() => undefined);

    const props = {
      backButton: false,
      title: "Test Title",
    };

    const { component } = setup(props);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render page info correctly", () => {
    const props = {
      backButton: true,
      title: "Test Title",
    };

    const { component } = setup(props);

    const backBtn = component.root.findAllByType(BackButton);
    const title = component.root.findByType("h3");

    expect(title.children).toContain(props.title);
    expect(backBtn.length).toEqual(1);
  });

  it("should render navigation for logged in user", () => {
    useAuthStatus.mockImplementationOnce(() => "1");

    const props = {
      backButton: false,
      title: "Test Title",
    };

    const { component } = setup(props);

    const settings = component.root.findAllByType(SettingsDropdown);
    const account = component.root.findAllByType(AccountDropdown);

    expect(settings.length).toEqual(1);
    expect(account.length).toEqual(1);
  });

  it("should render navigation for unauthenticated user", () => {
    useAuthStatus.mockImplementationOnce(() => undefined);

    const props = {
      backButton: false,
      title: "Test Title",
    };

    const { component } = setup(props);

    const account = component.root.findAllByType(AccountDropdown);

    expect(account.length).toEqual(0);
  });
});
