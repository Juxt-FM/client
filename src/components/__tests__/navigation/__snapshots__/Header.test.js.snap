// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render for logged in user without crashing 1`] = `
<header>
  <div>
    <h3>
      Test Title
    </h3>
  </div>
  <ul>
    <li>
      <a
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </li>
    <li>
      <div>
        <a
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-ellipsis-h fa-w-16 "
            data-icon="ellipsis-h"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </a>
      </div>
    </li>
    <li>
      <div>
        <a
          onClick={[Function]}
        >
          <img
            alt="logged in user"
            src="https://www.pererasys.com/_next/static/images/me-0e70979b96b772f832a278693ee0cd0e.jpg"
          />
        </a>
      </div>
    </li>
  </ul>
</header>
`;

exports[`<Header /> should render for unauthenticated user without crashing 1`] = `
<header>
  <div>
    <h3>
      Test Title
    </h3>
  </div>
  <ul>
    <li>
      <a
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </li>
    <li>
      <div>
        <a
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-ellipsis-h fa-w-16 "
            data-icon="ellipsis-h"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </a>
      </div>
    </li>
    <li />
  </ul>
</header>
`;
